orbit/pkg/token/reader.go

```diff
@@ -0,0 +1,76 @@
+package token
+
+import (
+	"os"
+	"sync"
+	"time"
+)
+
+// Reader is used to read the token value from a file
+type Reader struct {
+	Path string // a path to a file containing a token
+
+	mu     sync.Mutex // ensures atomic reads/writes; protects the following fields
+	cached string     // the value of the token, a.k.a. the contents of the file
+	mtime  time.Time  // the mtime of the file when the token was last read
+}
+
+// Read returns the token value from the file only if the file is
+// expired or the cached value is empty
+func (r *Reader) Read() (string, error) {
+	changed, err := r.HasChanged()
+	if err != nil {
+		return "", err
+	}
+
+	if changed || r.GetCached() == "" {
+		if err := r.readFile(); err != nil {
+			return "", err
+		}
+	}
+
+	return r.GetCached(), nil
+}
+
+// HasChanged checks if the in-memory `value` has changed by comparing
+// the chached `r.mtime` value with the `mtime` of the file at
+// `r.path`
+func (r *Reader) HasChanged() (bool, error) {
+	info, err := os.Stat(r.Path)
+	if err != nil {
+		return false, err
+	}
+	mtime := info.ModTime()
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return !mtime.Equal(r.mtime), nil
+}
+
+// HasExpired checks if 1 hour has passed since the last recorded `mtime` of
+// the token file
+func (r *Reader) HasExpired() bool {
+	return time.Now().After(r.mtime.Add(1 * time.Hour))
```

Comment:
Should you try to acquire the lock before performing this check here?

Reply:
great catch, thanks!